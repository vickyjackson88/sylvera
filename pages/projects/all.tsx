import Link from "next/link";

import styles from "../../styles/AllProjects.module.css";

interface ProjectsProps {
  projects: string[];
}

export const AllProjects = ({ projects }: ProjectsProps) => {
  return (
    <div className={styles.container}>
      <h1 className={styles.title}>Live Projects</h1>
      {projects.map((project) => (
        <div key={project} className={styles.card}>
          <Link
            data-testid={`${project}-link`}
            href={`/projects/${project}`}
            className={styles.link}
          >
            <span>{project.replace("_", " ")}</span>
            <span>â†’</span>
          </Link>
        </div>
      ))}
    </div>
  );
};

// Using server-side rendering with cache control headers
// instead of static rendering because the response changes
export const getServerSideProps = async ({ res }: any) => {
  res.setHeader(
    'Cache-Control',
    'public, s-maxage=10, stale-while-revalidate=60'
  )
  const response = await fetch("https://pm25.lass-net.org/API-1.0.0/project/all/");
  const responseText = await response.text();
  const projects = responseText.split(/\r?\n/).filter((project) => project);
  return {
    props: { projects }, // will be passed to the page component as props
  };
};

export default AllProjects;
