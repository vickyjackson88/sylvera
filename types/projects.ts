export type Feed = {
  device_id: string;
  s_t0: number;
  s_h0: number;
  s_d0: number;
  gps_lat: number;
  gps_lon: number;
  timestamp: string;
};

export type Project = {
  source: string;
  version: string;
  num_of_records: number;
  descriptions: {
    URL: string;
  };
  feeds: Feed[];
};
