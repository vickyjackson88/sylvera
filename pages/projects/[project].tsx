import { Project as ProjectType } from "../../types/projects";
import styles from "../../styles/Project.module.css";
import Link from "next/link";

interface ProjectProps {
  project: ProjectType;
}

const getFeedItems = (feed: any) => {
  const feedItems = [
    {
      label: "Device ID",
      data: feed.device_id,
    },
    {
      label: "Latitude",
      data: feed.gps_lat,
    },
    {
      label: "Longitude",
      data: feed.gps_lon,
    },
    {
      label: "Time of entry",
      data: `at ${new Date(feed.timestamp).toLocaleTimeString()} on ${new Date(
        feed.timestamp
      ).toLocaleDateString()}`,
    },
  ];

  return feedItems.map((feedItem) => {
    return (
      <div key={feedItem.label} className={styles.feedItemContent}>
        <span>{feedItem.label}</span>
        <span className={styles.data}>{feedItem.data}</span>
      </div>
    );
  });
};

export const Project = ({ project }: ProjectProps) => {
  const url = new URL(window.location.href);
  const projectName = url.pathname.split("/")[2];
  return (
    <div className={styles.container}>
      <div className={styles.titleContainer}>
        <Link href={"/projects/all"} className={styles.link}>
          ‚Üê
        </Link>
        <h1 className={styles.title}>{projectName}</h1>
      </div>
      <div>{`Number of feeds: ${project.num_of_records}`}</div>
      <div>{`Source: ${project.source}`}</div>
      <div>{`Version: ${project.version}`}</div>
      {project.num_of_records >= 10 && (
        <div className={styles.feed}>
          <h2 className={styles.subtitle}>Top 10 feeds:</h2>
          {project.feeds.map((feed, index) => {
            return (
              index < 10 && (
                <div
                  className={styles.feedItem}
                  data-testid="feed-item"
                  key={feed.device_id}
                >
                  {getFeedItems(feed)}
                </div>
              )
            );
          })}
        </div>
      )}
    </div>
  );
};

export const getServerSideProps = async ({ query, res }: any) => {
  res.setHeader(
    'Cache-Control',
    'public, s-maxage=10, stale-while-revalidate=60'
  )
  const response = await fetch(
    `https://pm25.lass-net.org/API-1.0.0/project/${query.project}/latest/`
  );
  const project = await response.json();
  return {
    props: { project }, // will be passed to the component as props
  };
};

export default Project;
