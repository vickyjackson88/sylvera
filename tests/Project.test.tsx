import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect'

import Project from "../pages/projects/[project]";
import { Project as ProjectType } from "../types/projects";

let mockProject: ProjectType = {
  source: "test source",
  version: "2022-11-26T19:46:11Z",
  num_of_records: 99,
  descriptions: {
    URL: "",
  },
  feeds: [
    {
      device_id: "test device id",
      s_t0: 1,
      s_h0: 2,
      s_d0: 3,
      gps_lat: 4,
      gps_lon: 5,
      timestamp: "2022-11-26T23:12:22Z",
    },
  ],
};

describe("Snapshot tests: projects/all", () => {
  it("renders Project DOM unchanged", () => {
    const { container } = render(<Project project={mockProject} />);
    expect(container).toMatchSnapshot();
  });
});

describe("Component tests: projects/[project]", () => {
  it("renders no feeds if there are less than 10", () => {
    const mockProjectWithUnder10Feeds = {
      ...mockProject,
      num_of_records: 9,
    };
    render(<Project project={mockProjectWithUnder10Feeds} />);
    const feedItem = screen.queryByTestId("feed-item");
    expect(feedItem).not.toBeInTheDocument();
  });

  it("renders feeds if there are 10", async () => {
    const mockProjectWith10Feeds = {
      ...mockProject,
      num_of_records: 10,
    };
    render(<Project project={mockProjectWith10Feeds} />);
    const feedItem = await screen.findByTestId("feed-item");
    expect(feedItem).toBeInTheDocument();
  });

  it("renders feeds if there are more than 10", async () => {
    const mockProjectWith10Feeds = {
      ...mockProject,
      num_of_records: 11,
    };
    render(<Project project={mockProjectWith10Feeds} />);
    const feedItem = await screen.findByTestId("feed-item");
    expect(feedItem).toBeInTheDocument();
  });
});
