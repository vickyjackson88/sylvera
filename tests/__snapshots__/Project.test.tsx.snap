// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests: projects/all renders Project DOM unchanged 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="titleContainer"
    >
      <a
        class="link"
        href="/projects/all"
      >
        ‚Üê
      </a>
      <h1
        class="title"
      />
    </div>
    <div>
      Number of feeds: 99
    </div>
    <div>
      Source: test source
    </div>
    <div>
      Verison: 2022-11-26T19:46:11Z
    </div>
    <div
      class="feed"
    >
      <h2
        class="subtitle"
      >
        Top 10 feeds:
      </h2>
      <div
        class="feedItem"
        data-testid="feed-item"
      >
        <div
          class="feedItemContent"
        >
          <span>
            Device ID
          </span>
          <span
            class="data"
          >
            test device id
          </span>
        </div>
        <div
          class="feedItemContent"
        >
          <span>
            Latitude
          </span>
          <span
            class="data"
          >
            4
          </span>
        </div>
        <div
          class="feedItemContent"
        >
          <span>
            Longitude
          </span>
          <span
            class="data"
          >
            5
          </span>
        </div>
        <div
          class="feedItemContent"
        >
          <span>
            Time of entry
          </span>
          <span
            class="data"
          >
            at 23:12:22 on 26/11/2022
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
